hare3/algorithm_test.go

```diff
@@ -0,0 +1,36 @@
+package hare3
+
+import (
+	"crypto/rand"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+)
+
+var (
+	id1 = randHash20()
+	id2 = randHash20()
+	id3 = randHash20()
+
+	val1 = randHash20()
+	val2 = randHash20()
+	val3 = randHash20()
+
+	msgHash1 = randHash20()
+
+	values1 = []types.Hash20{val1}
+	values2 = []types.Hash20{val1, val2}
+	values3 = []types.Hash20{val1, val2, val3}
+
+	rPre = NewAbsRound(0, -1)
+	r0   = NewAbsRound(0, 0)
+	r1   = NewAbsRound(0, 1)
+	r2   = NewAbsRound(0, 2)
+	r3   = NewAbsRound(0, 3)
+	r4   = NewAbsRound(0, 4)
```

Comment:
let's avoid global state for tests. if you want to reuse same setup for tests that can be accomplished with simple object

Reply:
I not sure I would consider this global state, since these are not intended to be written, just read. Unfortunately due to the types involved they can't be made consts. Could you elaborate on your suggested sue of a simple object, I'm not sure exactly what you are suggesting?

Reply:
really poor implementation of this pattern is in spacemesh api modules. in your case it is smaller but it is still not good.
for example if i need more than 3 ids, will i add more here? same for values1/2/3

i think better pattern is to write stateful tester class that will generate and manage collections of objects. so for example in this simple case it would be

```go
type hareTester struct {
    ids []int
    values []types.Hash20
}

func (t hareTester) getValues(positions ...int) []types.Hash32 {
    // and here it just selects values using positions, maybe even extends values is position is not found
}
```

Reply:
Hi @dshulyak sorry for the delay in getting back to you here. Is your concern that we have global state or is it that we'll need to keep adding values?

Reply:
the pattern with global state setup is a pain for next person that will work on this module

Reply:
so you are saying that you will be able to test everything with 3 ids and 3 different set of values? can you check API module, does testing there look ok to you?